### ISSUE TITLE
After sync failure, item fetching hangs for all subscriptions

### ISSUE THREAD

**Issue by mrtcode** [zotero/arkivo#6]
After sync failure, item fetching hangs for all subscriptions. After this error API accepts new subscriptions, even successfully  subscribes to them, but never responds to them (client timeouts). Probably after the thrown error some asynchronous operation doesn't resolve/reject promise or doesn't callback.

Invalid subscription key is enough for this problem to happen.
```
arkivo:controller [sync#19] [zwusubgufr] failed to sync subscription: Invalid key +655ms
arkivo:trace Error: Invalid key
```

**Comment by mrtcode** [comment]
I'll try to find the problem.

**Comment by inukshuk** [comment]
So `sync` jobs come in but are never processed? I would guess that means there is a [stuck active job](https://github.com/Automattic/kue#error-handling) (or that the worker crashed completely). (You can view the active jobs if you open the web API in a browser)

**Comment by mrtcode** [comment]
@inukshuk so it seems it's not related to job processing. This is `zotero-api-node` and `stream-server` problem. Is there only one stream-server connection per Arkivo process? The problem is that stream-server closes the connection when creating a subscription with an invalid key, or something else happens what stream-server don't like. And Arkivo isn't aware about lost stream server connection and just do nothing.

@dstillman this is related to the discussed about recovering from stream-server connection errors. But this is also a stream-server and zotero-api-node incompatibility problem.

**Comment by inukshuk** [comment]
That's odd, because Arkivo should definitely be sending regular pings to detect connection loss (TCP TTL was not enough because of proxies silently dropping connections etc.). You can [configure]( https://github.com/zotero/arkivo/blob/master/config/default.json#L15) the ping/pong intervals, but there's probably something -- are the pings being sent at all?

**Comment by mrtcode** [comment]
So when the key is invalid Stream.closed is called with CLOSE_BAD_KEY (4403), and when stream-server closes a connection (because restarting), Stream.closed is called with CLOSE_NORMAL (1000).

This happens here:
https://github.com/mrtcode/zotero-api-node/blob/d7c7bd98fb43fcc6406e7b4066d054624441b67c/lib/stream.js#L226-L226

In both cases it seems that no retry is set and ping intervals are cleared.

**Comment by inukshuk** [comment]
This looks like a logical error on my part. Since Arkivo handles multiple subscription it's not very sensible to abort if the key of a single subscription is invalid, so we should probably not close the connection at all in that case. Restart seems to be a special case, too: maybe we should add a code for CLOSE_RESTART to tell clients that they can re-connect again.

@dstillman what do you think?

**Comment by dstillman** [comment]
Yeah, this is mostly a problem in stream-server itself. A multi-subscription connection shouldn't be closed on various errors. We'll fix that to send most errors as messages instead, and then we can update zotero-api-node to pass those along.

**Comment by mrtcode** [comment]
So if a connection with stream-server is lost and/or it doesn't send `subscriptionsCreated` response, Arkivo HTTP API blocks the client request (when creating subscription) forever until the `subscriptionsCreated` is received. We should probably fail those client requests if stream-server is unreachable. Especially if we expect to use a webhook plugin and the client will be Zotero website. Otherwise client has to timeout.

@dstillman what do you think about the part that client should immediately get an error response if stream-server is unavailable?

That was also the reason why Arkivo API never ever respond after stream-server initiated socket close, when zotero-api-node doesn't try to reconnect. It's just `subscriptionsCreated` even is not triggered in zotero-api-node/lib/stream.js which doesn't trigger Arkivo/lib/listeners.js, and therefore doesn't respond to client.

@inukshuk maybe you have an advice how to return an error event from stream.js to the client in the current Arkivo architecture? Error can be just a closed socket, a connection error, and also we will probably add an additional event type `error` to return errors from stream-server instead of closing socket. The catch is that stream-server can return an error without any reference to the subscriptions that a client is trying to create. Therefore we don't know to which of the waiting clients to respond the error.

It needs a key/topic to find to which client requests to respond?
https://github.com/mrtcode/arkivo/blob/aeb25587f752d3c3582598516f61e0eccca0ee61/lib/listener.js#L83-L83

**Comment by dstillman** [comment]
Yeah, I noticed while working on the client that not knowing what request a response was associated with made things a little tricky. Errors aside, it means you can't even have clean asynchronous code, because you don't know when a send() has finished or whether it was successful (e.g., you can't determine when a `createSubscriptions` finished without checking the payload of all `subscriptionsCreated` responses).

This isn't not really an issue in the client at the moment, but it seems like it'd definitely be an issue when Arkivo needs to return a status code for an HTTP request, but that request depends on a request that it makes to stream-server that might fail.

I guess one option would be an optional request ID that could be passed by the client and that would be returned in responses, both successful and not. (If the connection was closed, you wouldn't know what caused it, but we're planning to stop doing that.)

(The astute reader will notice that this is a problem that regular HTTP solves. My first implementation of stream-server actually used Server-Sent Events and HTTP POSTs, but I [rewrote it](https://github.com/zotero/stream-server/commit/e2496f19e3b2896ee7f1aed8caa79a09ac48f9e5) when I realized that that would make horizontal scaling much trickier, at least without adding another dependency like Redis or some other message broker. So now we get to reimplement a request/response model on top of WebSockets.)

**Comment by mrtcode** [comment]
Yeah, passing an identifier in a stream-server request and returning it in the response is the right way to do it, I think. Because this is asynchronous protocol - there can be many subscription creations initiated simultaneously and many listeners waiting for the response for the exact request. In a synchronous protocol this would be simple - request, response, request, response, etc.

By using request IDs in Akrivo we will be able to reliably track each stream-server request.

**Comment by mrtcode** [comment]
Another problem is that subscribing to many different keys/topics is problematic. This relates to stream-server too. So let's say there are 100 subscriptions with different keys. One key is removed in dataserver. Then stream-server is restarted. After reconnect, arkivo tries to resubscribe to the same subscriptions with one request, but one key fails because of the wrong API key, therefore whole request fails. 

Before https://github.com/zotero/stream-server/pull/8 the failure looked as a closed connection. With this PR the connection is no longer closed, and just returns an error event. But still the same problem applies - one wrong API key interrupts whole re-subscription.

Actually the error not necessary must be a wrong API key, it can just be any dataserver non-success response code.

That means stream-server shouldn't fail if one API key request failed, and instead continue processing other subscriptions in the request. But a request subscription limit must be set to prevent internal DOS to dataserver if i.e. 1 million invalid key subscription request would be sent.

**Comment by mrtcode** [comment]
**UPDATE: fixed**
Also doesn't resubscribe if the previous resubscription process failed. stream-server, zotero-api-node and arkivo in this test are all in master branch.

```
  arkivo:listener [f0ga6vscjz] listening for updates of /users/1... +0ms
  zotero:stream websocket closed unexpectedly, retry in 10000ms +31s
  zotero:stream opening websocket to http://localhost:18081... +10s
  zotero:stream websocket opened successfully +9ms
  zotero:stream sending "createSubscriptions" message +0ms
  zotero:stream {"action":"createSubscriptions","subscriptions":[{"apiKey":"rGWO8xlHCtkEeShmrpDFlzfJ","topics":["/users/1"]}]} +0ms
  zotero:stream "connected" message received +2ms
  zotero:stream {"event":"connected","retry":10000} +0ms
  zotero:stream websocket closed unexpectedly, retry in 10000ms +10s
  zotero:stream opening websocket to http://localhost:18081... +10s
  zotero:stream websocket opened successfully +4ms
  zotero:stream "connected" message received +1ms
  zotero:stream {"event":"connected","retry":10000} +0ms <----- FINISHES HERE AND DOESN'T RESUBSCRIBE
```

**Comment by mrtcode** [comment]
A few more ideas to think about:

1. What happens when Arkivo tries to resubscribe to 1000 topics and 500 of them fails because some of the requests from stream-server to dataserver failed because of a network problem or a random dataserver problem. Are we going to live with another 500? and after another resubscribe, they would suddenly come back?  Are we going to detect that failure, close connection and start again?
2. Now about missing updates when disconnected. After reconnect, arkivo, will no longer knows if topics were updated or not while arkivo was disconnected. Therefore would be good to resynchronize items.
3. The same is with startup. Listeners are only attached after synchronization? if dataserver response takes 0.2s, and there is one worker, and we have 1000 subscriptions, it will take more than 3 minutes to synchronize all subscriptions. but if there are many items to synchronize, it can take much longer. arkivo listener subscribes to stream-server for the topic updates only after this process, it means the earlier the item was synchronizer the higher probability that some updates are missed in the time span between synchronization and subscription to stream-server. Maybe it would be possible to firstly subscribe to stream-server - **modify stream-server to return library version on subscription creation** - and than act on that library version - fetch items if missing. But I'm not sure if this is possible, have to understand better how arkivo works.

Or maybe I missed something.
